### DISCUSSION TITLE
API rules that allow me to list all others organizer and participants in a Algorithm contest

### DISCUSSION THREAD

**Discussion by alaireselene** [pocketbase/pocketbase#4654] [Q&A]
Hello guys, I'm building a website that use PB for storing contests and submissions.
My PB has 3 collections: `users`, `contests` and `submissions`
- `users` have default fields from PB, plus `type` field which is enum: `default`, `organizer`, `admin`
- `contests` have `organizers` and `participants` relation field that relate to `users` collection
- `submissions` have `contest` relate to `contests` and `user` relate to `users`
Now I want to write API rules that:
- For `contests`: Organizers and participants can list all contests that they work in, and see all other organizers and participants in those contests like them. Only users with `admin` type can create and delete, `admin` + `organizer` can modify
- For `submissions`: Organizers that in the contest `contest` can list and view all submissions in that contest. Participant can create submissions.

I tested by login with `Zeus` user info.

It strange that when I let all rules blank, PB just let me list contests, but when I access to one contest as an `organizer`, it show nothing in `organizers` field, and show nothing in `participants`. And I cannot list any submission in selected contest.

Finally, I've tried to write list rule in `contests` like that: `organizers.id ?= @request.auth.id || participants.id ?= @request.auth.id`, it only show the data of me in `organizers` field, and show nothing in `participants`.

`contests` collection looks like this:
![image](https://github.com/pocketbase/pocketbase/assets/56793922/e637a095-7d44-4c99-85c4-1f9cdc5020f6)

And data of `USACO Section 5` from query:
![image](https://github.com/pocketbase/pocketbase/assets/56793922/de7f82d9-580d-4c49-aeac-c3f260d03d7d)

I'm totally new in PB and confused to solve this problem. Can someone please explain to me what's wrong and how to fix it? Thank you so much!

**Comment by alaireselene** [comment]
Additionally, my JS code only use `expand` and `fields`, so there are no more filters.

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

Your list rule for the `contests` collection looks correct for me.
Are you sure that you are properly processing the json response?

>  it only show the data of me in organizers field, and show nothing in participants.

I don't understand what do you mean by this. If you mean that not all expanded fields are being populated then make sure that your client request satisfy the View API rule of the collections you are expanding (in your case that would be the `users` collection). If you have a View API rule like: `id = @request.auth.id` then you'll be able to expand only your own auth record no matter whether it is part of the `organizers` or `participants` field.

If the above doesn't help, please export your collections configuration as json (_Admin UI > Settings > Export collections_) and try to elaborate more clearly what requests you are sending and what is the expected outcome.

**Reply by alaireselene** [reply]
Oh, thank you for point out the problem in `users` collection, I remove that rule and it works like a charm. Can you suggest me the right way to write `users` View API rule in my case, so that I can learn how to write rule for other collections?

**Reply by ganigeorgiev** [reply]
I'm not sure that I'll able to help you as I'm still not quite sure that I understand your collections structure, nor what is the desired behavior.
Based on your title, if you want a List and View `users` API rule that will allow list/view access to users participating in the same contest as the current authenticated one (no matter whether they are organizers or participants), then something like the below should work:

```js
// require to be authenticated
@request.auth.id != "" &&
(
    // check if it is the current authenticated user
    id = @request.auth.id ||
    // or require the user we want to list/view to be part in one of the current authenticated user contents
    (
        // join the contests for the current authenticated users
        (@collection.contests.organizers.id ?= @request.auth.id || @collection.contests.participants.id ?= @request.auth.id) &&
        // require the user we want to list/view to be either an organizer or participants in one of those contests that we previously joined
        (@collection.contests.organizers.id ?= id || @collection.contests.participants.id ?= id)
    )
)
```

**Reply by ganigeorgiev** [reply]
Actually the above rule is not entirely correct because a single field cannot be 2 different values at the same time - `organizers.id` cannot match both `id` and `@request.auth.id` (aka. similar problem described in https://github.com/pocketbase/pocketbase/discussions/4319).
So one possible workaround could be to join the second `@collection.contests` with an [:alias](https://pocketbase.io/docs/api-rules-and-filters/#filters-syntax) or maybe more simpler - reference the field using [back-relation](https://pocketbase.io/docs/working-with-relations/#back-relations). Something like:

```js
@request.auth.id != "" &&
(
    id = @request.auth.id ||
    @request.auth.contests_via_organizers.organizers.id ?= id ||
    @request.auth.contests_via_organizers.participants.id ?= id ||
    @request.auth.contests_via_participants.organizers.id ?= id ||
    @request.auth.contests_via_participants.participants.id ?= id
)
```

_edit: updated to include both organizers and participants options_

**Reply by alaireselene** [reply]
Mmm, seems like I need to work around this problem more. I will post my solution soon, thank you so much for your suggestions!
